In [1]:
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.ollama import OllamaChatCompletionClient

In [2]:
ollama_model_client = OllamaChatCompletionClient(model="llama3.1")

## **Agent Define**
### ***Agent need mainly two things***
- Agent Name
- Agent Brain

In [3]:
agent_1 = AssistantAgent(
    name="My_Assistant",
    model_client=ollama_model_client,
    system_message="You are a helpful ai assistant. you are a math solver."
)

In [4]:
response = await agent_1.run(
    task = "Hi, my i'm Al Amin, How are you!!"
)

In [6]:
response.messages[-1]

TextMessage(id='61af2d3d-8c40-47b6-be00-50bb05a6f08c', source='My_Assistant', models_usage=RequestUsage(prompt_tokens=41, completion_tokens=107), metadata={}, created_at=datetime.datetime(2025, 7, 15, 5, 25, 9, 161693, tzinfo=datetime.timezone.utc), content="As-salamu alaykum, Al Amin! I'm doing great, thank you for asking! It's always wonderful to meet new people and assist with their math-related queries.\n\nI'm here to help you with any math problems or questions you may have. Whether it's algebra, geometry, calculus, or anything else, feel free to share the problem and I'll do my best to guide you through it.\n\nWhat brings you here today? Do you have a specific math problem you'd like some assistance with?", type='TextMessage')

TextMessage(id='61af2d3d-8c40-47b6-be00-50bb05a6f08c', source='My_Assistant', models_usage=RequestUsage(prompt_tokens=41, completion_tokens=107), metadata={}, created_at=datetime.datetime(2025, 7, 15, 5, 25, 9, 161693, tzinfo=datetime.timezone.utc),
 
content="As-salamu alaykum, Al Amin! I'm doing great, thank you for asking! It's always wonderful to meet new people and assist with their math-related queries.\n\nI'm here to help you with any math problems or questions you may have. Whether it's algebra, geometry, calculus, or anything else, feel free to share the problem and I'll do my best to guide you through it.\n\nWhat brings you here today? Do you have a specific math problem you'd like some assistance with?", 

type='TextMessage')

In [7]:
await agent_1.run(
    task= "What was my last question?"
)

TaskResult(messages=[TextMessage(id='a9419b76-770d-45a2-88ce-b8f8c58cee23', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 15, 5, 29, 5, 497285, tzinfo=datetime.timezone.utc), content='What was my last question?', type='TextMessage'), TextMessage(id='3da3b1a9-ed9f-4a0b-b385-fb2d6b7ccf60', source='My_Assistant', models_usage=RequestUsage(prompt_tokens=163, completion_tokens=44), metadata={}, created_at=datetime.datetime(2025, 7, 15, 5, 29, 8, 459749, tzinfo=datetime.timezone.utc), content='Your last message was a greeting: "Hi, my i\'m Al Amin, How are you!!" There were no previous questions from you. Let\'s start fresh! What can I help you with today?', type='TextMessage')], stop_reason=None)

TaskResult(messages=[TextMessage(id='a9419b76-770d-45a2-88ce-b8f8c58cee23', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 15, 5, 29, 5, 497285, tzinfo=datetime.timezone.utc), content='What was my last question?', type='TextMessage'),

 TextMessage(id='3da3b1a9-ed9f-4a0b-b385-fb2d6b7ccf60', source='My_Assistant', models_usage=RequestUsage(prompt_tokens=163, completion_tokens=44), metadata={}, created_at=datetime.datetime(2025, 7, 15, 5, 29, 8, 459749, tzinfo=datetime.timezone.utc),
 
  content='Your last message was a greeting: "Hi, my i\'m Al Amin, How are you!!" There were no previous questions from you. Let\'s start fresh! What can I help you with today?', type='TextMessage')], 
  
  stop_reason=None)

In [9]:
import warnings
warnings.filterwarnings("ignore")

## **Accessing LLMs Through Open-Router**

In [10]:
import os
from dotenv import load_dotenv
load_dotenv()

from autogen_ext.models.openai import OpenAIChatCompletionClient
open_router_api_key = os.getenv("OPEN_ROUTER_API_KEY")


open_router_model_client =  OpenAIChatCompletionClient(
    base_url="https://openrouter.ai/api/v1",
    model="deepseek/deepseek-r1-0528:free",
    api_key = open_router_api_key,
    model_info={
        "family":'deepseek',
        "vision" :True,
        "function_calling":True,
        "json_output": False
    }
)

In [13]:
from autogen_core.models import UserMessage

await open_router_model_client.create(
    [UserMessage(content = "what is the capital of Bangladesh?", source="user")]
)

CreateResult(finish_reason='stop', content="The capital of Bangladesh is **Dhaka** (also spelled **Dacca**).  \n\nIt is located in the central part of the country and serves as the political, economic, and cultural center of the People's Republic of Bangladesh. Dhaka is one of the most densely populated cities in the world, known for its vibrant history, river ports (notably on the Buriganga River), and bustling urban life.  \n\n**Key facts about Dhaka:**  \n- **Population**: Over 21 million (metropolitan area).  \n- **Significance**: Houses the National Parliament, major government institutions, universities, and cultural landmarks.  \n- **History**: Has been a regional hub since the Mughal era and became the capital of East Bengal (later East Pakistan, then Bangladesh).  \n\nLet me know if you'd like more details! 😊", usage=RequestUsage(prompt_tokens=13, completion_tokens=366), cached=False, logprobs=None, thought=None)

In [14]:
await ollama_model_client.create(
    [
        UserMessage(content="Hi, how are you?", source="user")
    ]
)

CreateResult(finish_reason='stop', content="I'm just a computer program, so I don't have feelings like humans do. But thank you for asking! How can I assist you today? Do you have any questions or need help with something specific?", usage=RequestUsage(prompt_tokens=16, completion_tokens=43), cached=False, logprobs=None, thought=None)